File: android/src/main/java/com/example/bluetooth_print_plus/bluetooth_print_plus/TscCommandPlugin.java
package com.example.bluetooth_print_plus.bluetooth_print_plus;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import androidx.annotation.NonNull;

import com.gprinter.command.LabelCommand;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.RequestPermissionsResultListener;


public class TscCommandPlugin implements FlutterPlugin, MethodCallHandler, RequestPermissionsResultListener {
    public MethodChannel channel;
    private static final String TAG = "BluetoothPrintPlusPlugin-TSC";

    private final LabelCommand tscCommand = new LabelCommand();


    public void setUpChannel(MethodChannel channel) {
        this.channel = channel;
        this.channel.setMethodCallHandler(this);
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        channel = new MethodChannel(binding.getBinaryMessenger(), "bluetooth_print_plus_tsc");
        channel.setMethodCallHandler(this);
    }


    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        Integer x = call.argument("x");
        Integer y = call.argument("y");
        String content = call.argument("content");
        Integer width = call.argument("width");
        Integer height = call.argument("height");
        Integer rotation = call.argument("rotation");
        String method = call.method;

        switch (method) {
            case "selfTest":
                this.tscCommand.addSelfTest();
                result.success(true);
                break;
            case "cleanCommand":
                this.tscCommand.clrCommand();
                result.success(true);
                break;
            case "getCommand":
                Object[] elements = this.tscCommand.getCommand().toArray();
                byte[] datas = new byte[elements.length];
                for (int i = 0; i < elements.length; i++) {
                    Byte item = (Byte) elements[i];
                    datas[i] = item;
                }
                result.success(datas);
                break;
            case "print":
                Integer copies = call.argument("copies");
                assert copies != null;
                this.tscCommand.addPrint(1, copies);
                result.success(true);
                break;
            case "gap":
                Integer gap = call.argument("gap");
                assert gap != null;
                this.tscCommand.addGap(gap);
                result.success(true);
                break;
            case "speed":
                Integer speed = call.argument("speed");
                assert speed != null;
                this.tscCommand.addSpeed(LabelCommand.SPEED.valueOf("SPEED" + speed));
                result.success(true);
                break;
            case "density":
                Integer density = call.argument("density");
                assert density != null;
                this.tscCommand.addDensity(LabelCommand.DENSITY.valueOf("DNESITY" + density));
                result.success(true);
                break;
            case "cls":
                this.tscCommand.addCls();
                result.success(true);
                break;
            case "size":
                assert width != null;
                assert height != null;
                this.tscCommand.addSize(width, height);
                result.success(true);
                break;
            case "text":
                Integer xMulti = call.argument("xMulti");
                Integer yMulti = call.argument("yMulti");
                assert x != null;
                assert y != null;
                this.tscCommand.addText(
                        x, y,
                        LabelCommand.FONTTYPE.SIMPLIFIED_24_CHINESE,
                        LabelCommand.ROTATION.valueOf("ROTATION_" + rotation),
                        LabelCommand.FONTMUL.valueOf( "MUL_" + xMulti), LabelCommand.FONTMUL.valueOf("MUL_" + yMulti), content
                );
                result.success(true);
                break;
            case "image":
                byte[] bytes = call.argument("image");
                assert bytes != null;
                assert x != null;
                assert y != null;
                Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
                this.tscCommand.addBitmap(x, y, LabelCommand.BITMAP_MODE.OVERWRITE, bitmap.getWidth(), bitmap);
                result.success(true);
                break;
            case "barCode":
                String codeType = call.argument("codeType");
                Boolean readable = call.argument("readable");
                Integer narrow = call.argument("narrow");
                Integer wide = call.argument("wide");
                assert x != null;
                assert y != null;
                assert height != null;
                assert narrow != null;
                assert wide != null;
                this.tscCommand.add1DBarcode(
                    x, y,
                    getCodeType(codeType),
                    height,
                    Boolean.TRUE.equals(readable) ? LabelCommand.READABEL.EANBEL : LabelCommand.READABEL.DISABLE,
                    LabelCommand.ROTATION.valueOf("ROTATION_" + rotation),
                    narrow, wide, content
                );
                result.success(true);
                break;
            case "qrCode":
                Integer cellWidth = call.argument("cellWidth");
                assert x != null;
                assert y != null;
                assert cellWidth != null;
                this.tscCommand.addQRCode(
                    x, y,
                    LabelCommand.EEC.LEVEL_M,
                    cellWidth,
                    LabelCommand.ROTATION.valueOf("ROTATION_" + rotation),
                    content
                );
                result.success(true);
                break;
            case "bar":
                assert x != null;
                assert y != null;
                assert width != null;
                assert height != null;
                this.tscCommand.addBar(x, y, width, height);
                result.success(true);
                break;
            case "box":
                Integer endX = call.argument("endX");
                Integer endY = call.argument("endY");
                Integer linThickness = call.argument("linThickness");
                assert x != null;
                assert y != null;
                assert endX != null;
                assert endY != null;
                assert linThickness != null;
                this.tscCommand.addBox(
                    x, y,
                    endX, endY,
                    linThickness
                );
                result.success(true);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    @Override
    public boolean onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        return false;
    }

    public LabelCommand.BARCODETYPE getCodeType(String codeType) {
        switch (codeType) {
            case "39":
                return LabelCommand.BARCODETYPE.CODE39;
            case "93":
                return LabelCommand.BARCODETYPE.CODE93;
            case "ITF":
                return LabelCommand.BARCODETYPE.ITF14;
            case "UPCA":
                return LabelCommand.BARCODETYPE.UPCA;
            case "UPCE":
                return LabelCommand.BARCODETYPE.UPCE;
            case "CODABAR":
                return LabelCommand.BARCODETYPE.CODABAR;
            case "EAN8":
                return LabelCommand.BARCODETYPE.EAN8;
            case "EAN13":
                return LabelCommand.BARCODETYPE.EAN13;
            default:
                return LabelCommand.BARCODETYPE.CODE128;
        }
    }
}

File: ios/Classes/TscCommandPlugin.m
//
//  TscCommand.m
//  bluetooth_print_plus
//
//  Created by amoLink on 2024/4/27.
//

#import "TscCommandPlugin.h"
#import "TscCommand.h"

@interface TscCommandPlugin ()

@property (nonatomic, strong) TscCommand *tscCommand;

@end

@implementation TscCommandPlugin
+ (void)registerWithRegistrar:(nonnull NSObject<FlutterPluginRegistrar> *)registrar {}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSMutableDictionary *argumentsDict = call.arguments;
    // NSLog(@"tsc call.method : %@", call.method);
    int x = [[argumentsDict valueForKey:@"x"] intValue];
    int y = [[argumentsDict valueForKey:@"y"] intValue];
    int width = [[argumentsDict valueForKey:@"width"] intValue];
    int height = [[argumentsDict valueForKey:@"height"] intValue];
    int rotation = [[argumentsDict valueForKey:@"rotation"] intValue];
    NSString *content = [argumentsDict valueForKey:@"content"];
    
    if ([@"selfTest" isEqualToString:call.method]) {
        [self.tscCommand addSelfTest];
        result(@(YES));
    } else if ([@"cleanCommand" isEqualToString:call.method]) {
        self.tscCommand = nil;
        result(@(YES));
    } else if ([@"getCommand" isEqualToString:call.method]) {
        result([self.tscCommand getCommand]);
    } else if ([@"print" isEqualToString:call.method]) {
        int copies = [[argumentsDict valueForKey:@"copies"] intValue];
        [self.tscCommand addPrint:1 :copies];
        result(@(YES));
    } else if ([@"gap" isEqualToString:call.method]) {
        int gap = [[argumentsDict valueForKey:@"gap"] intValue];
        [self.tscCommand addGapWithM:gap withN:0];
        result(@(YES));
    } else if ([@"speed" isEqualToString:call.method]) {
        int speed = [[argumentsDict valueForKey:@"speed"] intValue];
        [self.tscCommand addSpeed:speed];
        result(@(YES));
    } else if ([@"density" isEqualToString:call.method]) {
        int density = [[argumentsDict valueForKey:@"density"] intValue];
        [self.tscCommand addSpeed:density];
        result(@(YES));
    } else if ([@"cls" isEqualToString:call.method]) {
        [self.tscCommand addCls];
        result(@(YES));
    } else if ([@"size" isEqualToString:call.method]) {
        [self.tscCommand addSize:width :height];
        result(@(YES));
    } else if ([@"text" isEqualToString:call.method]) {
        int xMulti = [[argumentsDict valueForKey:@"xMulti"] intValue];
        int yMulti = [[argumentsDict valueForKey:@"yMulti"] intValue];
        [self.tscCommand addTextwithX:x withY:y withFont:@"TSS24.BF2" withRotation:rotation withXscal:xMulti withYscal:yMulti withText:content];
        result(@(YES));
    } else if ([@"image" isEqualToString:call.method]) {
        FlutterStandardTypedData *imageData = call.arguments[@"image"];
        UIImage *image = [UIImage imageWithData:imageData.data];
        
        [self.tscCommand addBitmapwithX:x withY:y withMode:0  withImage:image];
        result(@(YES));
    } else if ([@"barCode" isEqualToString:call.method]) {
        NSString *codeType = [argumentsDict valueForKey:@"codeType"];
        BOOL readable = [[argumentsDict valueForKey:@"readable"] boolValue];
        int narrow = [[argumentsDict valueForKey:@"narrow"] intValue];
        int wide = [[argumentsDict valueForKey:@"wide"] intValue];
        [self.tscCommand add1DBarcode:x :y :codeType :height :readable :rotation :narrow :wide :content];
        result(@(YES));
    } else if ([@"qrCode" isEqualToString:call.method]) {
        int cellWidth = [[argumentsDict valueForKey:@"cellWidth"] intValue];
        [self.tscCommand addQRCode:x :y :@"M" :cellWidth :@"A" :rotation :content];
        result(@(YES));
    }
    else if ([@"bar" isEqualToString:call.method]) {
        [self.tscCommand addBar:x :y :width :height];
        result(@(YES));
    }
    else if ([@"box" isEqualToString:call.method]) {
        int endX = [[argumentsDict valueForKey:@"endX"] intValue];
        int endY = [[argumentsDict valueForKey:@"endY"] intValue];
        int linThickness = [[argumentsDict valueForKey:@"linThickness"] intValue];
        [self.tscCommand addBox:x :y :endX :endY :linThickness];
        result(@(YES));
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}

- (TscCommand *)tscCommand {
    if (!_tscCommand) {
        _tscCommand = [TscCommand new];
    }
    return _tscCommand;
}


@end
